// @payloadcms/plugin-cloud/dist/plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "payloadCloud", {
    enumerable: true,
    get: function() {
        return payloadCloud;
    }
});
const _email = require("./email");
const _afterDelete = require("./hooks/afterDelete");
const _beforeChange = require("./hooks/beforeChange");
const _uploadCache = require("./hooks/uploadCache");
const _staticHandler = require("./staticHandler");
const _webpack = require("./webpack");
const payloadCloud = (pluginOptions)=>(incomingConfig)=>{
        let config = {
            ...incomingConfig
        };
        const webpack = (0, _webpack.extendWebpackConfig)(incomingConfig);
        config.admin = {
            ...config.admin || {},
            webpack
        };
        if (process.env.PAYLOAD_CLOUD !== 'true') {
            return config // only modified webpack
            ;
        }
        const cachingEnabled = pluginOptions?.uploadCaching !== false && !!process.env.PAYLOAD_CLOUD_CACHE_KEY;
        const apiEndpoint = pluginOptions?.endpoint || 'https://cloud-api.payloadcms.com';
        // Configure cloud storage
        if (pluginOptions?.storage !== false) {
            config = {
                ...config,
                collections: (config.collections || []).map((collection)=>{
                    if (collection.upload) {
                        return {
                            ...collection,
                            hooks: {
                                ...collection.hooks || {},
                                afterChange: [
                                    ...collection.hooks?.afterChange || [],
                                    ...cachingEnabled ? [
                                        (0, _uploadCache.getCacheUploadsAfterChangeHook)({
                                            endpoint: apiEndpoint
                                        })
                                    ] : []
                                ],
                                afterDelete: [
                                    ...collection.hooks?.afterDelete || [],
                                    (0, _afterDelete.getAfterDeleteHook)({
                                        collection
                                    }),
                                    ...cachingEnabled ? [
                                        (0, _uploadCache.getCacheUploadsAfterDeleteHook)({
                                            endpoint: apiEndpoint
                                        })
                                    ] : []
                                ],
                                beforeChange: [
                                    ...collection.hooks?.beforeChange || [],
                                    (0, _beforeChange.getBeforeChangeHook)({
                                        collection
                                    })
                                ]
                            },
                            upload: {
                                ...typeof collection.upload === 'object' ? collection.upload : {},
                                disableLocalStorage: true,
                                handlers: [
                                    ...typeof collection.upload === 'object' && Array.isArray(collection.upload.handlers) ? collection.upload.handlers : [],
                                    (0, _staticHandler.getStaticHandler)({
                                        cachingOptions: pluginOptions?.uploadCaching,
                                        collection
                                    })
                                ]
                            }
                        };
                    }
                    return collection;
                }),
                upload: {
                    ...config.upload || {},
                    useTempFiles: true
                }
            };
        }
        // Configure cloud email
        const apiKey = process.env.PAYLOAD_CLOUD_EMAIL_API_KEY;
        const defaultDomain = process.env.PAYLOAD_CLOUD_DEFAULT_DOMAIN;
        if (pluginOptions?.email !== false && apiKey && defaultDomain) {
            config.email = (0, _email.payloadCloudEmail)({
                apiKey,
                config,
                defaultDomain
            });
        }
        return config;
    };
