// @payloadcms/plugin-nested-docs/dist/fields/breadcrumbs.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const createBreadcrumbsField = (relationTo, overrides = {})=>({
        name: 'breadcrumbs',
        type: 'array',
        localized: true,
        fields: [
            {
                name: 'doc',
                type: 'relationship',
                relationTo,
                maxDepth: 0,
                admin: {
                    disabled: true
                }
            },
            {
                type: 'row',
                fields: [
                    {
                        name: 'url',
                        label: 'URL',
                        type: 'text',
                        admin: {
                            width: '50%'
                        }
                    },
                    {
                        name: 'label',
                        type: 'text',
                        admin: {
                            width: '50%'
                        }
                    }
                ]
            },
            ...overrides?.fields || []
        ],
        admin: {
            readOnly: true,
            ...overrides?.admin || {}
        },
        ...overrides || {}
    });
const _default = createBreadcrumbsField;
