// @payloadcms/plugin-nested-docs/dist/hooks/resaveChildren.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _populateBreadcrumbs = /*#__PURE__*/ _interop_require_default(require("../utilities/populateBreadcrumbs"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const resaveChildren = (pluginConfig, collection)=>async ({ doc, req: { locale, payload }, req })=>{
        const resaveChildrenAsync = async ()=>{
            const children = await payload.find({
                collection: collection.slug,
                depth: 0,
                locale,
                req,
                where: {
                    parent: {
                        equals: doc.id
                    }
                }
            });
            try {
                children.docs.forEach(async (child)=>{
                    const updateAsDraft = typeof collection.versions === 'object' && collection.versions.drafts && child._status !== 'published';
                    await payload.update({
                        id: child.id,
                        collection: collection.slug,
                        data: {
                            ...child,
                            breadcrumbs: await (0, _populateBreadcrumbs.default)(req, pluginConfig, collection, child)
                        },
                        depth: 0,
                        draft: updateAsDraft,
                        locale,
                        req
                    });
                });
            } catch (err) {
                payload.logger.error(`Nested Docs plugin has had an error while re-saving a child document.`);
                payload.logger.error(err);
            }
        };
        await resaveChildrenAsync();
        return undefined;
    };
const _default = resaveChildren;
