// @payloadcms/plugin-cloud/dist/utilities/authAsCognitoUser.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "authAsCognitoUser", {
    enumerable: true,
    get: function() {
        return authAsCognitoUser;
    }
});
const _amazoncognitoidentityjs = require("amazon-cognito-identity-js");
let sessionAndToken = null;
const authAsCognitoUser = async (username, password)=>{
    // TODO: Check that isValid evaluates expiration
    if (sessionAndToken?.isValid()) {
        return sessionAndToken;
    }
    const userPool = new _amazoncognitoidentityjs.CognitoUserPool({
        ClientId: process.env.PAYLOAD_CLOUD_COGNITO_USER_POOL_CLIENT_ID,
        UserPoolId: process.env.PAYLOAD_CLOUD_COGNITO_USER_POOL_ID
    });
    const authenticationDetails = new _amazoncognitoidentityjs.AuthenticationDetails({
        Password: password,
        Username: username
    });
    const cognitoUser = new _amazoncognitoidentityjs.CognitoUser({
        Pool: userPool,
        Username: username
    });
    const result = await new Promise((resolve, reject)=>{
        cognitoUser.authenticateUser(authenticationDetails, {
            onFailure: (err)=>{
                reject(err);
            },
            onSuccess: (res)=>{
                resolve(res);
            }
        });
    });
    sessionAndToken = result;
    return sessionAndToken;
};
