// @payloadcms/plugin-cloud/dist/utilities/getStorageClient.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "getStorageClient", {
    enumerable: true,
    get: function() {
        return getStorageClient;
    }
});
const _clientcognitoidentity = require("@aws-sdk/client-cognito-identity");
const _clients3 = /*#__PURE__*/ _interop_require_wildcard(require("@aws-sdk/client-s3"));
const _credentialproviders = require("@aws-sdk/credential-providers");
const _authAsCognitoUser = require("./authAsCognitoUser");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {};
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
let storageClient = null;
let session = null;
let identityID;
const getStorageClient = async ()=>{
    if (storageClient && session?.isValid()) {
        return {
            identityID,
            storageClient
        };
    }
    session = await (0, _authAsCognitoUser.authAsCognitoUser)(process.env.PAYLOAD_CLOUD_PROJECT_ID, process.env.PAYLOAD_CLOUD_COGNITO_PASSWORD);
    const cognitoIdentity = new _clientcognitoidentity.CognitoIdentityClient({
        credentials: (0, _credentialproviders.fromCognitoIdentityPool)({
            clientConfig: {
                region: 'us-east-1'
            },
            identityPoolId: process.env.PAYLOAD_CLOUD_COGNITO_IDENTITY_POOL_ID,
            logins: {
                [`cognito-idp.us-east-1.amazonaws.com/${process.env.PAYLOAD_CLOUD_COGNITO_USER_POOL_ID}`]: session.getIdToken().getJwtToken()
            }
        })
    });
    const credentials = await cognitoIdentity.config.credentials();
    // @ts-expect-error
    identityID = credentials.identityId;
    storageClient = new _clients3.S3({
        credentials,
        region: process.env.PAYLOAD_CLOUD_BUCKET_REGION
    });
    return {
        identityID,
        storageClient
    };
};
